### DISCUSSION TITLE
Polling the page source in a thread - looking for unexpected pop-ups

### DISCUSSION THREAD

**Discussion by cliffgibb** [appium/appium#16562] [Q&A] [answered]
I have developed an approach to handle unexpected pop-ups so that they do not interrupt the test flow. For example, a permissions pop-up, a presidential alert, an amber alert, app promotion, etc. For these known pop-ups, I want them to be programmatically dismissed, then allow the test flow to continue.

The approach has been to continuously retrieve the pageSource from the appium driver in a separate thread (approx. every 700 ms). Once the the page source is retrieved, the text is used to compare to all of our known pop-up definitions. If one is found then it is handled and the test may continue.

For Android, the getPageSource is super fast, and this approach has been working for some time.

For iOS, the process takes longer. I have reduced the retrieval time by using the recommended approach of using `executeScript("mobile:source"...` and excluding the attributes not required:

```
pageSource = appiumDriver
    .executeScript("mobile:source",
                   ImmutableMap.of("format", "xml",
                                   "excludedAttributes", "visible,x,y,width,height,enabled,index,type,name")).toString();
```

The issue is that the iOS test scripts are taking double the time to complete, and our locators are timing out. So for now I have disabled this process so that tests may complete. 

It appears that while the "get page source" is running in a separate thread, there is a bottleneck with processing the Appium calls.

Has anyone attempted a similar approach? Any ideas how to poll the page source in the background without impacting the test script running at the same time?

**Comment by jlipps** [comment, marked as answer]
I would not do this. The Appium server queues commands anyway, so it is essentially single threaded in its operation. Also, polling every 700ms is probably too frequent, as you've noted, definitely slowing iOS down.

What I would probably do instead if you need this kind of functionality is wrap all your client side commands in an action that first checks for an alert. That way it only happens when you need to, and you know it's not going to happen right in the middle of doing something else, which could cause logic bugs.

I.e., (JS-style pseudocode), instead of this:

```js
// assume your popup handler is running in another thread
const el = await driver.findElement('xpath', '//foo')
await el.click()

// do more stuff
```

do something like this

```js
await noPopup(async () => {
    const el = await driver.findElement('xpath', '//foo')
    await el.click()
})
```

this makes your code quite ugly but it's safer and more performant. probably depending on your language there are interesting ways to handle this with e.g. macros or a DSL.

FWIW, most people choose to just not handle these types of cases since they're unpredictable and rare, and instead mark a test for retry if an unexpected popup occurs during it (which you could detect by looking for a popup in a failure teardown).
